src/main/java/com/oracle/nosql/spring/data/core/NosqlTemplateBase.java

```diff
@@ -178,6 +188,181 @@ protected boolean doCreateTableIfNotExists(
+                }
+            }
+
+            // extract entity details into maps
+            Map<String, FieldValue.Type> shardKeys = entityInformation.
+                    getShardKeys();
+            Map<String, FieldValue.Type> nonShardKeys = entityInformation.
+                    getNonShardKeys();
+
+            Map<String, String> entityShardMap = new LinkedHashMap<>();
+            shardKeys.forEach((k, v) -> entityShardMap.put(
+                    k.toLowerCase(), v.name().toLowerCase()));
+
+            Map<String, String> entityNonShardMap = new LinkedHashMap<>();
+            nonShardKeys.forEach((k, v) -> entityNonShardMap.put(
+                    k.toLowerCase(), v.name().toLowerCase()));
+
+            Map<String, String> entityOthersMap = new LinkedHashMap<>();
+            entityOthersMap.put(JSON_COLUMN.toLowerCase(), "json");
+
+            // convert maps to String. String format is {k1 v1, k2 v2 ...}
+            String tableShards = "{" + tableShardMap.entrySet().stream()
+                    .map(e -> e.getKey() + " " + e.getValue()).
+                    collect(Collectors.joining(",")) + "}";
+            String tableNonShards = "{" + tableNonShardMap.entrySet()
+                    .stream().map(e -> e.getKey() + " " + e.getValue()).
+                    collect(Collectors.joining(",")) + "}";
+            String tableOthers = "{" + tableOthersMap.entrySet().stream()
+                    .map(e -> e.getKey() + " " + e.getValue()).
+                    collect(Collectors.joining(",")) + "}";
+
+            String entityShards = "{" + entityShardMap.entrySet().stream()
+                    .map(e -> e.getKey() + " " + e.getValue()).
+                    collect(Collectors.joining(",")) + "}";
+            String entityNonShards = "{" + entityNonShardMap.entrySet()
+                    .stream().map(e -> e.getKey() + " " + e.getValue()).
+                    collect(Collectors.joining(",")) + "}";
+            String entityOthers = "{" + entityOthersMap.entrySet().stream()
+                    .map(e -> e.getKey() + " " + e.getValue()).
+                    collect(Collectors.joining(",")) + "}";
+
+            String msg;
+            // check shard keys and types match
+            if (!tableShards.equals(entityShards)) {
+                msg = String.format("Shard primary keys mismatch: " +
+                        "table=%s, entity=%s.", tableShards, entityShards);
+                errors.add(msg);
+            }
+
+            // check non-shard keys and types match
+            if (!tableNonShards.equals(entityNonShards)) {
+                msg = String.format("Non-shard primary keys mismatch: " +
+                                "table=%s, entity=%s.", tableNonShards,
+                        entityNonShards);
+                errors.add(msg);
+            }
+
+            // check non-primary keys and types match
+            if (!tableOthers.equals(entityOthers)) {
```

Comment:
IMO table can have other non-PK columns we shouldn't interfere with. All we should care about is kvjson column exists and has JSON type.

Reply:
No. Table associated with spring data can have only one non-PK and it should be kv_json_. Is it possible for spring to create a table with more than one non-PK?

Reply:
Spring SDK is not creating tables with non-PK cols other than kv_json_, but this doesn't mean the users can't alter the table.

Reply:
it is the case, I have a discussion with a customer that decided to add more columns used by another application using the Node.js driver. They don't want to modify the existing data

Reply:
Ok. modified to check only kv_json_ and it's type